File: apps/web/src/lib/categoryTree.test.ts
import { describe, it, expect } from 'vitest'
import {
  buildCategoryTree,
  categoryLabel,
  collectDescendantIds,
  flattenCategoryTree,
} from './categoryTree'
import type { Category } from '@/types'

function category(id: string, name: string, parentId: string | null = null): Category {
  return { id, name, type: 'out', parent_id: parentId, emoji: null }
}

describe('buildCategoryTree', () => {
  it('nests children under their parent, alphabetically at each level', () => {
    const roots = buildCategoryTree([
      category('transport', 'Transporte'),
      category('taxi', 'Taxi', 'transport'),
      category('bus', 'Bus', 'transport'),
      category('home', 'Hogar'),
    ])
    expect(roots.map((node) => node.category.name)).toEqual(['Hogar', 'Transporte'])
    expect(roots[1].children.map((node) => node.category.name)).toEqual(['Bus', 'Taxi'])
  })

  it('nests to any depth', () => {
    const flat = flattenCategoryTree([
      category('a', 'A'),
      category('b', 'B', 'a'),
      category('c', 'C', 'b'),
    ])
    expect(flat.map((entry) => [entry.category.name, entry.depth])).toEqual([
      ['A', 0],
      ['B', 1],
      ['C', 2],
    ])
  })

  it('promotes a child whose parent is missing to a root', () => {
    const roots = buildCategoryTree([category('orphan', 'Orphan', 'gone')])
    expect(roots.map((node) => node.category.name)).toEqual(['Orphan'])
  })

  it('keeps every category when the parent links form a cycle', () => {
    const flat = flattenCategoryTree([category('a', 'A', 'b'), category('b', 'B', 'a')])
    expect(flat).toHaveLength(2)
  })
})

describe('collectDescendantIds', () => {
  it('returns the category and everything below it', () => {
    const categories = [
      category('a', 'A'),
      category('b', 'B', 'a'),
      category('c', 'C', 'b'),
      category('d', 'D'),
    ]
    expect([...collectDescendantIds(categories, 'a')].sort()).toEqual(['a', 'b', 'c'])
    expect([...collectDescendantIds(categories, 'd')]).toEqual(['d'])
  })
})

describe('categoryLabel', () => {
  it('prefixes the emoji when there is one', () => {
    expect(categoryLabel({ ...category('t', 'Taxi'), emoji: '🚕' })).toBe('🚕 Taxi')
    expect(categoryLabel(category('t', 'Taxi'))).toBe('Taxi')
  })
})

File: apps/android/app/src/test/java/app/openlinks/spendtracker/CategoryTreeTest.kt
package app.openlinks.spendtracker

import app.openlinks.spendtracker.data.Category
import app.openlinks.spendtracker.ui.categoryChipLabel
import app.openlinks.spendtracker.ui.categoryLabel
import app.openlinks.spendtracker.ui.categoryPickerOptions
import app.openlinks.spendtracker.ui.flattenCategories
import org.junit.Assert.assertEquals
import org.junit.Test

private fun category(
    id: String,
    name: String,
    parentId: String? = null,
    emoji: String? = null,
) = Category(id = id, name = name, type = "out", parentId = parentId, emoji = emoji)

class CategoryTreeTest {
    private val categories = listOf(
        category("transport", "Transporte"),
        category("taxi", "Taxi", parentId = "transport", emoji = "🚕"),
        category("bus", "Bus", parentId = "transport"),
        category("home", "Hogar"),
    )

    @Test
    fun `orders children under their parent, alphabetically at each level`() {
        assertEquals(
            listOf("Hogar" to 0, "Transporte" to 0, "Bus" to 1, "Taxi" to 1),
            flattenCategories(categories).map { flat -> flat.category.name to flat.depth },
        )
    }

    @Test
    fun `nests to any depth`() {
        val deep = listOf(
            category("a", "A"),
            category("b", "B", parentId = "a"),
            category("c", "C", parentId = "b"),
        )
        assertEquals(listOf(0, 1, 2), flattenCategories(deep).map { flat -> flat.depth })
    }

    @Test
    fun `promotes a child whose parent is missing to a root`() {
        val orphan = listOf(category("orphan", "Orphan", parentId = "gone"))
        assertEquals(listOf("Orphan" to 0), flattenCategories(orphan).map { it.category.name to it.depth })
    }

    @Test
    fun `keeps every category when the parent links form a cycle`() {
        val looped = listOf(category("a", "A", parentId = "b"), category("b", "B", parentId = "a"))
        assertEquals(2, flattenCategories(looped).size)
    }

    @Test
    fun `picker options indent children and prefix the emoji`() {
        assertEquals(
            listOf(
                "home" to "Hogar",
                "transport" to "Transporte",
                "bus" to "    Bus",
                "taxi" to "    🚕 Taxi",
            ),
            categoryPickerOptions(categories),
        )
    }

    @Test
    fun `chip label marks a child instead of indenting it`() {
        val flattened = flattenCategories(categories).associateBy { flat -> flat.category.id }
        assertEquals("🚕 Taxi", categoryLabel(flattened.getValue("taxi").category))
        assertEquals("↳ 🚕 Taxi", categoryChipLabel(flattened.getValue("taxi")))
        assertEquals("Transporte", categoryChipLabel(flattened.getValue("transport")))
    }
}
